internal/graph/cached_resolver.go

```diff
@@ -0,0 +1,128 @@
+package graph
+
+import (
+	"bytes"
+	"context"
+	"encoding/base64"
+	"encoding/gob"
+	"fmt"
+	"time"
+
+	"github.com/karlseguin/ccache/v3"
+	"github.com/openfga/openfga/pkg/tuple"
+)
+
+const (
+	defaultMaxCacheSize = 10000
+	defaultCacheTTL     = 10 * time.Second
+)
+
+// CachedCheckResolver implements the CheckResolver interface in way that attempts to resolve
+// Check subproblems via prior computations before delegating the request to some underlying
+// CheckResolver.
+type CachedCheckResolver struct {
+	delegate     CheckResolver
+	cache        *ccache.Cache[*ResolveCheckResponse]
+	maxCacheSize int64
+	cacheTTL     time.Duration
+}
+
+var _ CheckResolver = (*CachedCheckResolver)(nil)
+
+// CachedCheckResolverOpt defines an option that can be used to change the behavior of cachedCeckResolver
+// instance.
+type CachedCheckResolverOpt func(*CachedCheckResolver)
+
+// WithMaxCacheSize sets the maximum size of the Check resolution cache. After this
+// maximum size is met, then cache keys will start being evicted with an LRU policy.
+func WithMaxCacheSize(size int64) CachedCheckResolverOpt {
+	return func(ccr *CachedCheckResolver) {
+		ccr.maxCacheSize = size
+	}
+}
+
+// WithCacheTTL sets the TTL (as a duration) for any single Check cache key value.
+func WithCacheTTL(ttl time.Duration) CachedCheckResolverOpt {
+	return func(ccr *CachedCheckResolver) {
+		ccr.cacheTTL = ttl
+	}
+}
+
+func WithExistingCache(cache *ccache.Cache[*ResolveCheckResponse]) CachedCheckResolverOpt {
+	return func(ccr *CachedCheckResolver) {
+		ccr.cache = cache
```

Comment:
Here, you should call `ccr.cache.Stop` before re-assigning

Reply:
Whatever way we decide, as long as we call `stop()` on the default cache if the user passed an existing cache, i'm happy :)

Reply:
Implemented the pattern as @jon-whit discussed above